sympy/tensor/array/expressions/tests/test_convert_array_to_indexed.py

```diff
@@ -44,3 +43,15 @@ def test_convert_array_to_indexed_main():
 
     assert convert_array_to_indexed(OneArray(3, 3), [i, j]) == 1
     assert convert_array_to_indexed(ZeroArray(3, 3), [i, j]) == 0
+
+    expr = Reshape(A, (27,))
+    assert convert_array_to_indexed(expr, [i]) == A[i // 9, i // 3 % 3, i % 3]
```

Comment:
Currently the c printer maps multiple indices to a single one:
```python
>>> from sympy import IndexedBase, Idx
>>> from sympy.printing.c import ccode
>>> i, j, k = Idx('i',3), Idx('j',3), Idx('k',3)
>>> A= IndexedBase('A')
>>> ccode(A[i,j,k])
A[9*i + 3*j + k]
```
Currently indices with integer arithmetic are not printed https://github.com/sympy/sympy/issues/13954

Reply:
If printing would work correctly, I'd assume that `A[i]` is printed given `ccode(A[i // 9, i // 3 % 3, i % 3])` (if we substitute manually in the above output).

Reply:
> Currently the c printer maps multiple indices to a single one

Maybe for the parsers at this point. Anyway, it's best to have these transformations in the array module rather than in the printing module.

Reply:
I do think this is generally a useful PR though, if only for the fact that it would allow to pass arbitrary `ArrayExpr`s to non-array-like printers and have it always print to something consistent

Reply:
> If printing would work correctly,

The example in the test is bad. You should rather create a one dimensional `ArraySymbol` and then reshape it into a multidimensional array (i.e. the opposite of what is shown). At this point you should get `A[9*i + 3*j + k]`.

Reply:
Would it make sense to also add tests (and code) that ensure that `convert_array_to_indexed(convert_indexed_to_array(obj)) == obj`? I realise it might not be possible to ensure this for all possible cases, but it sure would be nice. It's a bit hard to understand what a `Reshape` would be in terms of an `Indexed` (other than using multiple `Idx` in the same "index slot")

Reply:
There are already some cases like that in the tests (not in this PR). I don't believe `convert_indexed_to_array` would work for `Reshape` because it's pretty complicated to write a parser for all `Mod` and `floor` objects.